### PR TITLE
dev/core#1661 Allow phones with types longer than 16 chars to export

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Fixes a bug whereby attempting to export aa phone type field longer than 16 characters causes a crash on export

Before
----------------------------------------
Rename a phone type id to be more than 16 characters long, do an export including 'Phone type' in the export - crash

After
----------------------------------------
It's now possible to export.

Technical Details
----------------------------------------
This builds on some cleanup - notably https://github.com/civicrm/civicrm-core/pull/17951 & https://github.com/civicrm/civicrm-core/pull/17941 are not merged at time of writing and can be rebased out of this once merged.

The handling of the phone_type field is brought into line with other pseudoconstants whereby
- the pseudofield is added to the $fields array on the Query object
- the pseudofield is populated in convertToPseudoNames
In order to make the former work I had to pass the phone fields through the appendPseudoConstantsToFields function (we already do this on Contribution & some other BAO).
To make the second work I had to ensure that the location-type specific fields also get entries in the ```_pseudoConstantsSelect``` array

Comments
----------------------------------------
Replaces https://github.com/civicrm/civicrm-core/pull/16833 which failed tests

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
Test fail was not familiar but doesn't seem related CRM_Member_Form_MembershipTest::testFinancialEntiriesOnCancelledContribution

**Comment by eileenmcnaughton** [comment]
@demeritcowboy you tested the version of this that wasn't passing tests - maybe you can check this?

**Comment by demeritcowboy** [comment]
There was the case activity export. This doesn't sound familiar.

I see you've also done a replacement for the case one. I can check that?

**Comment by eileenmcnaughton** [comment]
@demeritcowboy yes - sorry I commented on the wrong PR

**Comment by eileenmcnaughton** [comment]
@pradpnayak or @yashodha I'm wondering if either you have any time to review this - there is a preliminary cleanup PR here https://github.com/civicrm/civicrm-core/pull/17951 which has just the first commit in it (if that is merged first I will rebase this)

File: CRM/Core/DAO/Phone.php [hunk 1]
```diff
@@ -289,7 +289,7 @@ public static function &fields() {
         'phone_type_id' => [
           'name' => 'phone_type_id',
           'type' => CRM_Utils_Type::T_INT,
-          'title' => ts('Phone Type'),
+          'title' => ts('Phone Type ID'),
```
Comment (colemanw) [inline CRM/Core/DAO/Phone.php:292]: Push-me pull-me again. I've been changing these fields to user-friendly names in the metadata (e.g. without the "ID"). That's best for form builder, search builder, and forms in general.

File: xml/schema/Core/Phone.xml [hunk 2]
```diff
@@ -148,7 +148,7 @@
   </field>
   <field>
     <name>phone_type_id</name>
-    <title>Phone Type</title>
+    <title>Phone Type ID</title>
```
Comment (colemanw) [inline xml/schema/Core/Phone.xml:151]: So I'm a -1 on this change and would prefer to override the label in `exportableFields` or wherever it needs to contain ID but leave it user-friendly in the metadata.

Whichever way we go we should be consistent with all fields, and at this point most titles in the metadata do *not* have the "ID" appended.

Comment (eileenmcnaughton) [reply xml/schema/Core/Phone.xml:151]: @colemanw so the fields are ID but the pseudoconstants all have names which make sense for the entity reference etc

Comment (eileenmcnaughton) [reply xml/schema/Core/Phone.xml:151]: So for example the fields array winds up with 2 fields - Phone Type - which has a label from the option_group.label & phone_type_id which is the integer. This also manifests in qf-search builder & I assume views filters on an integer not a  string when you add it in the url. We have got a few fields that are pretty confusing in search builder now as the labels got renamed so the titles are duplicated - but I think we should use the pseudoconstant titles & let the fields have the correct name

**Comment by eileenmcnaughton** [comment]
@colemanw I tried to split off the metadata part to this per ^^ but now I review what is happening there I think I'll have to change the tests twice if that merges first - if we get this merged I'll finish that one

**Comment by seamuslee001** [comment]
This seems consistent with the previous approach and noting that the title thing has been settled and field tittles will have ID and can be overridden in the html tittle for search purposes

**Pull request merged**